nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```
@@ -90,4 +132,141 @@ private static void invalidInstantiationError(
+      VariableTree varTree = (VariableTree) tree;
+      lhsTree = varTree.getType();
+      rhsTree = varTree.getInitializer();
+    } else {
+      AssignmentTree assignmentTree = (AssignmentTree) tree;
+      lhsTree = assignmentTree.getVariable();
+      rhsTree = assignmentTree.getExpression();
+    }
+    // rhsTree can be null for a VariableTree.  Also, we don't need to do a check
+    // if rhsTree is the null literal
+    if (rhsTree == null || rhsTree.getKind().equals(Tree.Kind.NULL_LITERAL)) {
+      return;
+    }
+    Type lhsType = ASTHelpers.getType(lhsTree);
+    Type rhsType = ASTHelpers.getType(rhsTree);
+    if (rhsTree instanceof NewClassTree
+        && ((NewClassTree) rhsTree).getIdentifier() instanceof ParameterizedTypeTree) {
+      // for the parameterized typed tree ASTHelpers.getType() returns a type that does not have
+      // annotations preserved
+      rhsType =
+          typeWithPreservedAnnotations(
+              (ParameterizedTypeTree) ((NewClassTree) rhsTree).getIdentifier());
+
+      ParameterizedTypeTree paramTypedTree =
+          (ParameterizedTypeTree) ((NewClassTree) rhsTree).getIdentifier();
+      // not generic
+      if (paramTypedTree.getTypeArguments().size() <= 0) {
+        return;
+      }
+    }
+    if (lhsType != null && rhsType != null) {
+      compareAnnotations(lhsType, rhsType, tree);
+    }
+  }
+
+  private void compareAnnotations(Type lhsType, Type rhsType, Tree tree) {
+    if (lhsType instanceof Type.ClassType && rhsType instanceof Type.ClassType) {
+      rhsType = supertypeMatchingLHS((Type.ClassType) lhsType, (Type.ClassType) rhsType);
+    }
+    List<Type> lhsTypeArguments = lhsType.getTypeArguments();
+    List<Type> rhsTypeArguments = rhsType.getTypeArguments();
+    // this error should already be generated by some other part of NullAway still adding a check to
+    // be safe
+    if (lhsTypeArguments.size() != rhsTypeArguments.size()) {
+      return;
+    }
```

We should throw a `RuntimeException` if this check fails rather than just returning; it's a violation of a fundamental assumption we are making